File: contests/abc245_py/src/abc245_g.py
import sys
import math
import heapq

N, M, K, L = map(int, input().split())
A = list(map(lambda x: int(x)-1, input().split()))
B = list(map(lambda x: int(x)-1, input().split()))
UVC = [list(map(int, input().split())) for _ in range(M)]
g = [[] for _ in range(N)]
for a, b, c in UVC:
    a -= 1
    b -= 1
    g[a].append((b, c))
    g[b].append((a, c))

def dijkstra(ninki):
    dist = [math.inf]*N
    appeared = [False]*N
    q = []
    for u in ninki:
        q.append((0, u))
        dist[u] = 0
    while q:
        d, u = heapq.heappop(q)
        if appeared[u]:
            continue
        appeared[u] = True
        for v, c in g[u]:
            if appeared[v]:
                continue
            d_ = d+c
            if d_ < dist[v]:
                dist[v] = d_
                heapq.heappush(q,d_, v))
    return dist


dists = {}  # (bitlength, 01) -> (nodeid -> cost)
for i in range(len(bin(K)[2:])+2):
    for flg in [0, 1]:
        # 該当する人気者を集める
        ninki = []
        for b in B:
            if ((A[b]>>i)&1)^flg:
                ninki.append(b)
        # dijkstra
        dists[(i, flg)] = dijkstra(ninki)
sys.exit(0)
anss = []
for u in range(N):
    ans = math.inf
    for i in range(len(bin(K)[2:])+2):
        flg = ((A[u]>>i)&1)
        ans = min(ans, dists[(i, flg)][u])
    anss.append(ans if ans != math.inf else -1)
print(*anss)

File: contests/abc245_rust/src/abc245_g.rs
use proconio::input;
use std::collections::BinaryHeap;
use std::cmp::min;
static INF: usize = std::usize::MAX;

fn print_vec<T: std::fmt::Display>(v: &[T]) {
    if v.len() == 0 {
        println!();
    } else {
        let mut v = v.iter();
        print!("{}", v.next().unwrap());
        for v in v {
            print!(" {}", v);
        }
        println!();
    }
}
fn main() {
    input! {
        N: usize,
        M: usize,
        K: usize,
        L: usize,
        A: [usize; N],
        B: [usize; L],
        UVC: [[usize; 3]; M],
        //m: usize,
        //a: [[i32]; n], // a is Vec<i32>, n-array.
        //ab: [[i32; n]; m] // `a` is Vec<Vec<i32>>, (m, n)-matrix.
    }
    let mut g: Vec<Vec<(usize, usize)>> = vec![Vec::new(); N];
    let mut dists: Vec<Vec<Vec<usize>>> = vec![vec![vec![INF; N]; 2]; 20];
    let mut anss: Vec<i64> = Vec::new();
    for uvc in UVC.iter() {
        let u = uvc[0];
        let v = uvc[1];
        let c = uvc[2];
        g[u-1].push((v-1, c));
        g[v-1].push((u-1, c));
    }
    fn dijkstra(ninki: Vec<usize>, g: &Vec<Vec<(usize, usize)>>, N: usize) -> Vec<usize>{
        let mut dist = vec![INF; N];
        let mut appeared: Vec<bool> = vec![false; N];
        let mut q: BinaryHeap<(usize, usize)> = BinaryHeap::new();
        for u in ninki {
            q.push((INF, u));
            dist[u] = 0;
        }
        while let Some((d_inv, u)) = q.pop() {
            let d = INF - d_inv;
            if appeared[u] {
                continue;
            }
            appeared[u] = true;
            for (v, c) in g[u].iter(){
                if appeared[*v]{
                    continue;
                }
                let d_ = d+(*c);
                if d_ < dist[*v]{
                    dist[*v] = d_;
                    q.push((INF-d_, *v));
                }
            }
        }
        return dist
    }

    for i in 0..20_usize{
        for flg in 0..2{
            let mut ninki:Vec<usize> = Vec::new();
            for u_ in B.iter(){
                let u = *u_-1;
                let a = A[u]-1;
                if ((a>>i)&1)^flg==1 {
                    ninki.push(u)
                }
            }
            dists[i][flg] = dijkstra(ninki, &g, N);
        }
    }
    for u in 0..N{
        let mut ans = INF;
        for i in 0..20_usize{
            let a = A[u]-1;
            let flg = (a>>i)&1;
            ans = min(ans, dists[i][flg][u]);
        }
        if ans==INF{
            anss.push(-1);
        } else {
            anss.push(ans as i64);
        }
    }
    println!("{}", anss.iter().map(|x| format!("{}", x)).collect::<Vec<_>>().join(" "))
}
